### PR TITLE
[Editor mocks] Validation on mock generation

### PR DESCRIPTION
**Pull request by arnaudlewis**
## Context

So far, we use to have a different format for mocks which corresponded to the format of the API.
We decided to move to the Editor format but it means that current users of SM will have to deal with incompatible mocks.
To solve that, we simple dynamically validate the mocks and if we detect that they are not there or simply invalid (common case will be the API format) then we regenerate fully the mocks.

### CODE REVIEW

File: packages/core/src/libraries/component.ts [hunk 1]
```diff
@@ -177,17 +166,24 @@ export function getComponentInfo(
...
   /* This illustrates the requirement for apps to pass paths to mocks */
-  const maybeMock = resolvePathsToMock({
-    paths: assetsPaths,
-    from,
-    sliceName,
-  });
+  const mocks = (() => {
+    const resolvedMocks = Files.readEntityFromFile<ComponentMocks>(
+      path.join(basePath, from, sliceName, "mocks.json"),
+      (payload: any) => {
```
Comment (MarcMcIntosh) [inline packages/core/src/libraries/component.ts]: linter warning here, maybe unknown would work better

File: packages/core/src/libraries/mocks.ts [hunk 2]
```diff
@@ -1,39 +0,0 @@
...
-  sliceName,
-}: {
-  path: string;
-  from: string;
-  sliceName: string;
-}): string {
-  return path.join(filePath, from, sliceName, "mocks.json");
-}
-
-export function resolvePathsToMock({
```
Comment (MarcMcIntosh) [inline packages/core/src/libraries/mocks.ts:17]: no-longer needed?

File: packages/core/src/models/LibrariesState.ts [hunk 3]
```diff
@@ -1,9 +1,10 @@
+import { SharedSliceContent } from "@prismicio/types-internal/lib/content/fields/slices/SharedSliceContent";
 import { SharedSlice } from "@prismicio/types-internal/lib/customtypes/widgets/slices";
 import * as t from "io-ts";
 // import { VariationMock } from "./Slice";
```
Comment (MarcMcIntosh) [inline packages/core/src/models/LibrariesState.ts]: nice, now the comments can be removed, (they where there to remind me about removing valiatio mock)

File: packages/slice-machine/lib/mock/Slice.ts [hunk 4]
```diff
@@ -61,19 +60,11 @@ export function buildSliceMockConfig(
...
+): ComponentMocks {
   const model = Slices.fromSM(smModel);
 
   const sliceMockConfig = buildSliceMockConfig(model, legacyMockConfig);
-  const sliceModel: SharedSlice = {
-    ...model,
-    variations: model.variations.map((v) => {
-      return {
-        ...v,
-        imageUrl: "",
```
Comment (MarcMcIntosh) [inline packages/slice-machine/lib/mock/Slice.ts:73]: why did we strip out the imageUrl before?

File: packages/slice-machine/lib/storybook.ts [hunk 5]
```diff
@@ -57,24 +59,22 @@ export default {
...
-    const mocks = Files.readFirstOf<
-      SharedSliceContent[],
-      Record<string, unknown>
-    >([customMocksPath, generatedMocksPath])((value: string) => {
-      return JSON.parse(value) as unknown as SharedSliceContent[];
-    });
-    if (!mocks) {
+    const mocks = Files.readEntity<ComponentMocks>(
+      mocksPath,
+      (payload: any) => {
```
Comment (MarcMcIntosh) [inline packages/slice-machine/lib/storybook.ts]: would unknown work?

File: packages/slice-machine/server/src/api/common/generate.ts [hunk 6]
```diff
@@ -20,14 +22,21 @@ export function generate(
     );
 
     components.forEach((c) => {
-      if (
-        !Files.exists(
```
Comment (MarcMcIntosh) [inline packages/slice-machine/server/src/api/common/generate.ts:24]: trying to think of what this was there for :/

File: packages/slice-machine/server/src/api/common/generate.ts [hunk 7]
```diff
@@ -20,14 +22,21 @@ export function generate(
...
+      const mocks = Files.readEntityFromFile<ComponentMocks>(
+        mocksPath,
+        (payload: unknown) => {
+          return getOrElseW(() => new Error("Invalid component mocks."))(
+            ComponentMocks.decode(payload)
+          );
+        }
+      );
+
+      if (!mocks || mocks instanceof Error) {
```
Comment (MarcMcIntosh) [inline packages/slice-machine/server/src/api/common/generate.ts]: ^ happens here now

File: packages/slice-machine/server/src/api/common/generate.ts [hunk 8]
```diff
@@ -20,14 +22,21 @@ export function generate(
...
+        const mocks: ComponentMocks = mock(
```
Comment (MarcMcIntosh) [inline packages/slice-machine/server/src/api/common/generate.ts:37]: is there no error for mocks being declared twice?

File: packages/slice-machine/tests/lib/mock/Slice.test.ts [hunk 9]
```diff
@@ -1,7 +1,7 @@
 import "@testing-library/jest-dom";
 import { WidgetTypes } from "@prismicio/types-internal/lib/customtypes/widgets";
 import { SlicesTypes } from "@prismicio/types-internal/lib/customtypes/widgets/slices";
-import { SliceSM, Slices } from "@slicemachine/core/build/models";
+import { SliceSM } from "@slicemachine/core/build/models";
 import { isRight } from "fp-ts/lib/Either";
 import MockSlice from "../../../lib/mock/Slice";
 // import allFieldSliceModel from "../../../tests/__mocks__/sliceModel";
```
Comment (MarcMcIntosh) [inline packages/slice-machine/tests/lib/mock/Slice.test.ts:7]: I'll need to add this test case back in

File: packages/slice-machine/tests/src/modules/screenshots/utils.ts [hunk 10]
```diff
@@ -28,7 +27,7 @@ export const getSelectedSliceDummyData = () => {
     extension: "js",
     model: dummyModel,
     screenshots: {},
-    mock: mocks as unknown as SharedSliceContent[],
+    mock: mocks as unknown as ComponentMocks,
```
Comment (MarcMcIntosh) [inline packages/slice-machine/tests/src/modules/screenshots/utils.ts:30]: this is where property based testing would come in handy :)

**Review by MarcMcIntosh — COMMENTED** [review]
Left a few comments :)

File: packages/core/src/libraries/mocks.ts (hunk 2 above)
Comment (arnaudlewis) [reply packages/core/src/libraries/mocks.ts:17]: no, I simplified the code, you'll see when reviewing, it's no longer necessary

File: packages/slice-machine/lib/mock/Slice.ts (hunk 4 above)
Comment (arnaudlewis) [reply packages/slice-machine/lib/mock/Slice.ts:73]: We use to require it, now we have a fallback system so if not specified, the validator will automatically set a default url that is used in the wroom already

File: packages/slice-machine/server/src/api/common/generate.ts (hunk 6 above)
Comment (arnaudlewis) [reply packages/slice-machine/server/src/api/common/generate.ts:24]: It used to check if we had existing mocks. I changed the logic to check 2 things:
- If we have existing mocks
- if the mocks are valid (using SharedSliceContent validator)

If one of these case is false, we regenerate

File: packages/slice-machine/server/src/api/common/generate.ts (hunk 7 above)
Comment (arnaudlewis) [reply packages/slice-machine/server/src/api/common/generate.ts]: yes that's the replacement for the condition above.

File: packages/slice-machine/server/src/api/common/generate.ts (hunk 8 above)
Comment (arnaudlewis) [reply packages/slice-machine/server/src/api/common/generate.ts:37]: oh yeah weird. I'll do some renaming around here.

File: packages/slice-machine/tests/lib/mock/Slice.test.ts (hunk 9 above)
Comment (arnaudlewis) [reply packages/slice-machine/tests/lib/mock/Slice.test.ts:7]: which case?

File: packages/slice-machine/tests/src/modules/screenshots/utils.ts (hunk 10 above)
Comment (arnaudlewis) [reply packages/slice-machine/tests/src/modules/screenshots/utils.ts:30]: but actually we could runtime validate the mocks here to type them properly and at least you're sure that's the right format for your tests maybe.
But yeah I guess your suggestion makes sense. 
That's something to consider for the future

File: packages/slice-machine/tests/lib/mock/Slice.test.ts (hunk 9 above)
Comment (MarcMcIntosh) [reply packages/slice-machine/tests/lib/mock/Slice.test.ts:7]: way down at the bottom, for allFieldSliceModel

File: cypress/integration/slices/00-create.spec.js [hunk 11]
```diff
@@ -57,29 +57,6 @@ describe("Create Slices", () => {
     );
     cy.readFile(type).should("contains", sliceName);
 
-    cy.readFile(pathToMock, "utf-8")
```
Comment (MarcMcIntosh) [inline cypress/integration/slices/00-create.spec.js:60]: what would be a good check to make sure the mocks was generated?

File: packages/core/src/node-utils/paths.ts [hunk 12]
```diff
@@ -91,9 +91,6 @@ export const CustomTypesPaths = (cwd: string): Paths =>
 export const CustomPaths = (cwd: string): Paths => Paths(cwd, "");
 export const PackagePaths = (cwd: string): Paths => Paths(cwd, "node_modules");
 export const SMConfigPath = (cwd: string): string => path.join(cwd, "sm.json");
-export const LibrariesStatePath = (cwd: string): string =>
```
Comment (MarcMcIntosh) [inline packages/core/src/node-utils/paths.ts:94]: I guess we should let user know that they can delete libraries-state

**Review by MarcMcIntosh — COMMENTED** [review]
Left a few questions :)

File: packages/slice-machine/tests/server/services/getEnv.test.js [hunk 13]
```diff
@@ -296,9 +297,13 @@ describe("getEnv", () => {
...
 
-    const { env } = await getEnv(TMP);
-    expect(env.applicationMode).toEqual(ApplicationMode.DEV);
-    expect(exitSpy).not.toHaveBeenCalled();
+    try {
+      const { env } = await getEnv(TMP);
+      expect(env.applicationMode).toEqual(ApplicationMode.DEV);
+      expect(exitSpy).not.toHaveBeenCalled();
+    } catch (e) {
+      console.error(e);
```
Comment (MarcMcIntosh) [inline packages/slice-machine/tests/server/services/getEnv.test.js]: why try/catch here?

File: packages/core/src/node-utils/paths.ts (hunk 12 above)
Comment (arnaudlewis) [reply packages/core/src/node-utils/paths.ts:94]: I guess so but also there will be remaining mocks in .slicemachine, the goal would be more to tell them that they can no longer versioned `.slicemachine` all together on GitHub and clean that up. Only the mock config is a problem there but we’ll probably no longer support it.
But for sure we could have a changelog to tell them what they could clean. If no mock config, just remove `.slicemachine` all together would be the best recommandation

File: packages/core/src/libraries/mocks.ts (hunk 2 above)
Comment (MarcMcIntosh) [reply packages/core/src/libraries/mocks.ts:17]: good

File: packages/core/src/node-utils/paths.ts (hunk 12 above)
Comment (MarcMcIntosh) [reply packages/core/src/node-utils/paths.ts:94]: cool, that'll need to go in the change-log when ever it's written :)

**Review by MarcMcIntosh — APPROVED** [review]

File: packages/slice-machine/tests/server/services/getEnv.test.js (hunk 13 above)
Comment (arnaudlewis) [reply packages/slice-machine/tests/server/services/getEnv.test.js]: debug stuff left here, I removed it

**Pull request merged**